### PR TITLE
Rewrite the LogStore component of raft (#177)

### LINKED ISSUE #177: Rewrite the LogStore component of raft
For the purpose of each interface, you can refer to the internal definition of `hashicorp/raft`, or refer to some existing implementation methods defined within `hashicorp/raft`.
```go
// raftLog implements raft.LogStore interface, we can use it to store raft logs
// how to store raft logs? we can use FlyDB/RocksDB/LevelDB/BoltDB to store raft logs
// maybe we can use FlyDB
type raftLog struct {
	// implement me
}

// newRaftLog returns a new raftLog
func newRaftLog() raft.LogStore {
	return &raftLog{}
}

func (r *raftLog) FirstIndex() (uint64, error) {
	//TODO implement me
	panic("implement me")
}

func (r *raftLog) LastIndex() (uint64, error) {
	//TODO implement me
	panic("implement me")
}

func (r *raftLog) GetLog(index uint64, log *raft.Log) error {
	//TODO implement me
	panic("implement me")
}

func (r *raftLog) StoreLog(log *raft.Log) error {
	//TODO implement me
	panic("implement me")
}

func (r *raftLog) StoreLogs(logs []*raft.Log) error {
	//TODO implement me
	panic("implement me")
}

func (r *raftLog) DeleteRange(min, max uint64) error {
	//TODO implement me
	panic("implement me")
}
```

### PR DESCRIPTION
**Pull request by saeid-a**
Raft's `LogStore` functions are implemented in two ways: `InMemory,` `BoltDB,` `RocksDB,` and `FlyDB.` The configuration of the implementation can be done through a text-based format.
closes #177

### CODE REVIEW

**Comment by sjcsjc123** [comment]
Very efficient! Perhaps we can try FlyDB as a raft log storage, after all, after benchmark testing, we found that FlyDB has better read and write performance than some common key value DBs!

**Comment by sjcsjc123** [comment]
However, a more common one is `rocksdb`, which may also require support for

**Comment by saeid-a** [comment]
@sjcsjc123 I've added `FlyDB` as well;

**Comment by saeid-a** [comment]
@sjcsjc123 added 'FlyDB' and `RocksDB` integration; please check.

**Comment by sjcsjc123** [comment]
Hello, our organization has sent you an invitation letter. Please check it carefully!

File: cluster/store/store_log.go [hunk 1]
```diff
@@ -1,45 +1,83 @@
...
+// Finally, it creates a new LogStore using the retrieved factory and an empty configuration map.
+// The created LogStore and an error (if any) are returned.
+func newRaftLog() (raft.LogStore, error) {
+	_ = Init()
 
-func (r *raftLog) StoreLog(log *raft.Log) error {
-	//TODO implement me
-	panic("implement me")
+	// Get the "memory" DataStoreFactory from the map
+	return getDataStore("memory", config.Options{})
```
Comment (sjcsjc123) [inline cluster/store/store_log.go:44]: You can adds fields in the `config/cluster_Options.go` to provide user defined storage methods, then passes them in somewhere such as meta/store/region, and then passes them to the `newRaft` method, replacing the "memory" here.

File: cluster/store/store_log.go [hunk 2]
```diff
@@ -1,45 +1,83 @@
...
 
-// raftLog implements raft.LogStore interface, we can use it to store raft logs
-// how to store raft logs? we can use FlyDB/RocksDB/LevelDB/BoltDB to store raft logs
-// maybe we can use FlyDB
-type raftLog struct {
-	// implement me
-}
+// DataStoreFactory is a function type that creates a new instance of a raft.LogStore.
+// It takes a configuration map as input and returns the created LogStore or an error if the creation fails.
+type DataStoreFactory func(conf config.Options) (raft.LogStore, error)
```
Comment (sjcsjc123) [inline cluster/store/store_log.go:12]: May be the `config.Options` should be replaced with Config in `config/cluster_options.go`?

File: cluster/store/store_log.go (hunk 1 above)
Comment (sjcsjc123) [reply cluster/store/store_log.go:44]: For example, you can store this field in the `store` in `func (s *store) newRaftNode()`, and then pass this field to the `newRaftLog` method. As for `config.Options{}`, you can remove it.

File: lib/datastore/boltdb.go [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+// init is a method on BoltDbStore that initializes the BoltDB store
+// It creates the necessary buckets if they don't exist
+func (ds *BoltDbStore) init() error {
+	ds.mux.RLock()
+	defer ds.mux.RUnlock()
+	tx, err := ds.conn.Begin(true)
+	if err != nil {
+		return err
+	}
+	defer check(tx.Rollback)
```
Comment (sjcsjc123) [inline lib/datastore/boltdb.go:58]: A great method for detecting errors!

File: lib/datastore/boltdb.go
Comment (sjcsjc123) [inline lib/datastore/boltdb.go:1]: The code is written very clearly and elegantly!

File: lib/datastore/rocksdb.go [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+type RocksDbStore struct {
+	mux  sync.RWMutex
+	conn *gorocksdb.DB
+}
+
+// NewLogRocksDbStorage is a function that creates a new RocksDB store
+// It takes a configuration map as input and returns a raft.LogStore and an error
+func NewLogRocksDbStorage(conf config.Options) (raft.LogStore, error) {
+	filename := conf.DirPath
+	options := gorocksdb.NewDefaultOptions()
```
Comment (sjcsjc123) [inline lib/datastore/rocksdb.go:22]: It seems that a method that supports user-defined rocksdb configuration can be added in the later stage, but there is currently no rush to implement it.

**Review by sjcsjc123 — COMMENTED** [review]
I have read all the code you wrote and overall it is very good. I have learned a lot from it. Thank you for your help!

File: cluster/store/store_log.go (hunk 1 above)
Comment (saeid-a) [reply cluster/store/store_log.go]: @sjcsjc123 I changed the option and config; please check again.

**Review by sjcsjc123 — APPROVED** [review]

**Pull request merged**